Mathlib/Analysis/Calculus/Gradient.lean

```diff
@@ -0,0 +1,472 @@
+  use δ; constructor
+  apply dpos
+  intro x' x1mem
+  have : ‖x - x'‖ ≤ δ:= by
+    rw [Metric.ball, Set.mem_setOf, dist_comm, dist_eq_norm] at x1mem
+    exact LT.lt.le x1mem
+  rw [Set.mem_setOf, norm_sub_rev x']
+  apply h x' this
+
+/- The equivalent statement of `HasFDerivAt` in the `ε - δ` language-/
+theorem HasFDeriv_iff_Convergence_Point {gradient' : (E →L[ℝ] ℝ)}:
+  HasFDerivAt f (gradient') x ↔ ∀ ε > (0 : ℝ), ∃ δ > (0 : ℝ), ∀ (x' : E),
+     ‖x - x'‖ ≤ δ → ‖f x' - f x - (gradient') (x' - x)‖ ≤ ε * ‖x - x'‖ := by
+  constructor
+  apply HasFDeriv_Convergence
+  apply Convergence_HasFDeriv
+
+theorem HasFDeriv_iff_Convergence :
+  HasFDerivAt f (f' x) x ↔ ∀ ε > (0 : ℝ), ∃ δ > (0 : ℝ), ∀ (x' : E),
+     ‖x - x'‖ ≤ δ → ‖f x' - f x - (f' x) (x' - x)‖ ≤ ε * ‖x - x'‖ := by
+  constructor
+  apply HasFDeriv_Convergence
+  apply Convergence_HasFDeriv
+
+/- The equivalent statement of `HasGradnAt` in the `ε - δ` language-/
+theorem HasGradnAt_iff_Convergence: HasGradnAt f gradf x ↔
+  ∀ ε > (0 : ℝ), ∃ δ > (0 : ℝ), ∀ (x' : E),
+  ‖x - x'‖ ≤ δ → ‖f x' - f x - ⟪gradf, (x' - x)⟫‖ ≤ ε * ‖x - x'‖ := by
+  rw [HasGradnAt, HasFDeriv_iff_Convergence_Point]
+  rfl
+
+/- The following theorems are about the translation of `HasFDerivAt` and `HasGradnAt`-/
+theorem HasGradn_HasFDeriv  (h: HasGradnAt f gradient x):
+  HasFDerivAt f (∇* gradient) x := h
+
+theorem HasFDeriv_HasGradn (h: HasFDerivAt f gradient' x): HasGradnAt f (grad gradient') x := by
+  rw [HasGradnAt, HasFDeriv_iff_Convergence_Point]
+  rw [HasFDeriv_iff_Convergence_Point] at h
+  intro e epos
+  rcases (h e epos) with ⟨d, dpos, h⟩
+  use d; constructor; exact dpos
+  intro x' xle; specialize h x' xle
+  rw [← toDual_of_toDual_eq_self]
+  exact h
+
+theorem HasFDeriv_HasGradn' (h: HasFDerivAt f (∇* gradient) x):
+   HasGradnAt f gradient x := by rw [HasGradnAt]; exact h
+
+theorem HasGradn_HasFDeriv' (h: HasGradnAt f (grad gradient') x): HasFDerivAt f gradient' x := by
+  rw [HasFDeriv_iff_Convergence_Point]
+  rw [HasGradnAt, HasFDeriv_iff_Convergence_Point ] at h
+  intro e epos
+  rcases (h e epos) with ⟨d, dpos, h⟩
+  use d; constructor; exact dpos
+  intro x' xle; specialize h x' xle
+  rw [← toDual_of_toDual_eq_self] at h
+  exact h
+
+/-
+### Calculations on Gradient
+-/
+variable {g : E → ℝ} {g' : E → (E →L[ℝ] ℝ)}
+variable (hf : HasGradnAt f gradf x) (hg : HasGradnAt g gradg x) (c : ℝ)
+
+theorem HasGradnAt.const_smul : HasGradnAt (fun x => c • f x) (c • gradf) x := by
+  let f':= ∇* gradf
+  have h₁: HasFDerivAt (fun x => c • f x) (c • f') x:= by
+    apply HasFDerivAt.const_smul
+    apply HasGradn_HasFDeriv hf
+  have h₂: c • f' = ∇* (c • gradf) := by
+    apply ContinuousLinearMap.ext_iff.mpr; intro x
+    have l₁ : (c • f') x = c * (f' x):= by rfl
+    have l₂ : (∇* (c • gradf)) x = ⟪c • gradf, x⟫ := by rfl
+    have l₃ : ⟪(c • gradf), x⟫ =  c * ⟪gradf, x⟫ := real_inner_smul_left gradf x c
+    rw[l₁, l₂, l₃]; rfl
+  rw[HasGradnAt, ← h₂]
+  exact h₁
```

Comment:
These should all follow immediately from the corresponding proofs for `HasFDerivAt`.  Example:
```suggestion
  rw [HasGradnAt] at hf ⊢
  convert hf.const_smul c
  simp
```